libp2p/protocols/kademlia/find.nim

```diff
@@ -7,115 +7,140 @@
+
+  return noNew
+
+proc dispatchFindNode*(
+    switch: Switch,
+    peer: PeerId,
+    key: Key,
+    addrs: Opt[seq[MultiAddress]] = Opt.none(seq[MultiAddress]),
 ): Future[Message] {.
     async: (raises: [CancelledError, DialFailedError, ValueError, LPStreamError])
 .} =
-  let conn = await kad.switch.dial(peerId, addrs, KadCodec)
+  let addrs = addrs.valueOr(switch.peerStore[AddressBook][peer])
+  let conn = await switch.dial(peer, addrs, KadCodec)
   defer:
     await conn.close()
 
-  let msg = Message(msgType: MessageType.findNode, key: targetId)
+  let msg = Message(msgType: MessageType.findNode, key: key)
   await conn.writeLp(msg.encode().buffer)
 
-  let reply = Message.decode(await conn.readLp(MaxMsgSize)).tryGet()
-  if reply.msgType != MessageType.findNode:
-    raise newException(ValueError, "Unexpected reply message type: " & $reply.msgType)
+  let reply = Message.decode(await conn.readLp(MaxMsgSize)).valueOr:
+    raise newException(ValueError, "FindNode reply decode fail")
 
   return reply
 
+proc updatePeers(kad: KadDHT, peerInfos: seq[PeerInfo]) {.raises: [].} =
+  for p in peerInfos:
+    discard kad.rtable.insert(p.peerId)
+    # Nodes might return different addresses for a peer, so we append instead of replacing
+    try:
+      var existingAddresses = kad.switch.peerStore[AddressBook][p.peerId].toHashSet()
+      for a in p.addrs:
+        existingAddresses.incl(a)
+      kad.switch.peerStore[AddressBook][p.peerId] = existingAddresses.toSeq()
+    except KeyError as exc:
+      debug "Could not update shortlist", err = exc.msg
+    # TODO: add TTL to peerstore, otherwise we can spam it with junk
+
 proc findNode*(
-    kad: KadDHT, targetId: Key
+    kad: KadDHT, key: Key
 ): Future[seq[PeerId]] {.async: (raises: [CancelledError]).} =
-  ## Iteratively search for the k closest peers to a target ID.
-
-  var initialPeers = kad.rtable.findClosestPeerIds(targetId, kad.config.replication)
-  var state = LookupState.init(
-    targetId, initialPeers, kad.config.alpha, kad.config.replication,
-    kad.rtable.config.hasher,
-  )
-  var addrTable: Table[PeerId, seq[MultiAddress]]
-  for p in initialPeers:
-    addrTable[p] = kad.switch.peerStore[AddressBook][p]
-
-  while not state.done:
-    let toQuery = state.selectAlphaPeers()
-    debug "Find node queries",
-      peersToQuery = toQuery.mapIt(it.shortLog()), addressTable = addrTable
-    var pendingFutures = initTable[PeerId, Future[Message]]()
+  ## Iteratively search for the k closest peers to a key.
+
+  var state = LookupState.init(kad, key)
 
-    for peer in toQuery.filterIt(kad.switch.peerInfo.peerId != it):
-      state.markPending(peer)
-      let addrs = addrTable.getOrDefault(peer, @[])
-      if addrs.len == 0:
-        state.markFailed(peer)
-        continue
-      pendingFutures[peer] =
-        kad.sendFindNode(peer, addrs, targetId).wait(chronos.seconds(5))
+  while not state.done():
+    let toQuery = state.selectAlphaPeers()
```

Comment:
```suggestion 
  while true:
    let toQuery = state.selectAlphaPeers()
    if toQuery.len == 0: break
```

`state.done()` will call `selectAlphaPeers` that is needed right below, better reuse that.